py/src/nupic/tests/encoders/date_test.py

```diff
@@ -0,0 +1,223 @@
+
+""" Unit tests for date encoder. """
+
+import datetime
+import numpy
+import unittest
+
+from nupic.encoders.date import DateEncoder
+
+"""
+def __init__(self,
+season=0,
+dayOfWeek=0,
+weekend=0,
+holiday=0,
+timeOfDay=0,
+customDays=0,
+holidays=((12, 25),)):
+"""
+
+class DateEncoderTest(unittest.TestCase):
+  """ Unit tests for DateEncoder class. """
+
+  def testDateEncoder(self):
+    """ Creating date encoder instance. """
+    # 3 bits for season, 1 bit for day of week, 1 for weekend, 5 for time of day
+    enc = DateEncoder(season=3, dayOfWeek=1, weekend=1, timeOfDay=5)
+    # In the middle of fall, Thursday, not a weekend, afternoon - 4th Nov,
+    # 2010, 14:55
+    d = datetime.datetime(2010, 11, 4, 14, 55)
+    # d = datetime.datetime(2010, 11, 1, 8, 55) # DEBUG
+    bits = enc.encode(d)
+    # Season is aaabbbcccddd (1 bit/month)
+    seasonExpected = [0,0,0,0,0,0,0,0,0,1,1,1]
+
+    # Week is MTWTFSS contrary to localtime documentation, Monday = 0 (for
+    # python datetime.datetime.timetuple()
+    dayOfWeekExpected = [0,0,0,1,0,0,0]
+
+    # Not a weekend, so it should be "False"
+    weekendExpected = [1, 0]
+
+    # Time of day has radius of 4 hours and w of 5 so each bit = 240/5
+    # min = 48min 14:55 is minute 14*60 + 55 = 895; 895/48 = bit 18.6
+    # should be 30 bits total (30 * 48 minutes = 24 hours)
+    timeOfDayExpected = (
+      [0,0,0,0,0,0,0,0,0,0,
+       0,0,0,0,0,0,1,1,1,1,
+       1,0,0,0,0,0,0,0,0,0])
+    expected = seasonExpected + dayOfWeekExpected + weekendExpected + timeOfDayExpected
+
+    print(enc.timeOfDayEncoder.size)
+    print(expected)
+    print(bits.dense.tolist())
+    self.assertEqual(bits.size, 51)
+    self.assertEqual(expected, bits.dense.tolist())
+
+
+
+  def testDayOfWeek(self):
+    """ Creating date encoder instance. """
+    # 1 bit for days in a week (x7 days -> 7 bits), no other fields encoded
+    enc = DateEncoder(dayOfWeek=1)
+    # In the middle of fall, Thursday, not a weekend, afternoon - 4th Nov,
+    # 2010, 14:55
+    d = datetime.datetime(2010, 11, 4, 14, 55)
+    # d = datetime.datetime(2010, 11, 1, 8, 55) # DEBUG
+    bits = enc.encode(d)
+
+    # Week is MTWTFSS contrary to localtime documentation, Monday = 0 (for
+    # python datetime.datetime.timetuple()
+    dayOfWeekExpected = [0,0,0,1,0,0,0] #Thu
+
+    expected = dayOfWeekExpected
+    self.assertEqual(bits.size, 7)
+    self.assertEqual(expected, bits.dense.tolist())
+
+
+  def testSeason(self):
+    """ Creating date encoder instance. """
+    # 3 bits for season (x4 seasons -> 12 bits), no other fields encoded
+    enc = DateEncoder(season=3)
+    # In the middle of fall, Thursday, not a weekend, afternoon - 4th Nov,
+    # 2010, 14:55
+    d = datetime.datetime(2010, 11, 4, 14, 55)
+    # d = datetime.datetime(2010, 11, 1, 8, 55) # DEBUG
+    bits = enc.encode(d)
+
+    # Season is aaabbbcccddd (1 bit/month)
+    seasonExpected = [0,0,0,0,0,0,0,0,0,1,1,1]
```

Comment:
this is failing with off by one shift: 
```
- [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1]
E     ?                                ---
E
E     + [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1]
E     ?  +++
```

But otherwise seems correct, do we want to figure, or just update the expected encoding?

Reply:
This is expected.  The scalar encoder used to "center" the activity, and now it does not.

This could be fixed by subtracting 1/2 of the radius from the input value and then applying modulus.  If this was done inside of the ScalarEncoder then it would exactly match the output of the original scalar encoder.

Reply:
> This could be fixed by subtracting 1/2 of the radius from the input value and then applying modulus. If this was done inside of the ScalarEncoder then it would exactly match the output of the original scalar encoder. 

Do we want to do that? Not needed, but probably a good idea to be "binary output compatible", as it'll make a lot of pity troubles with porting tests go away.

Reply:
Yes, I think we should do it, and do it in the ScalarEncoder, but I think it would require also doing the modulus in the ScalarEncoder, so it might not be as simple as it looks.

The modulus is an outstanding TODO anyways...

Reply:
is it ok if I just fix the test's expected data to pass now, to get this out of the plate for NAB? And I'll file an issue to make Scalar binary compatible, which would affect this encoder

Reply:
Of course.  Functionally, it should do the same thing so there is no issue.

The original Nupic had that exactl pattern of activity because it made the ScalarEncoder look nice for visualizations.

Reply:
> The original Nupic had that exactl pattern of activity because it made the ScalarEncoder look nice for visualizations. 

speaking of which, RDSE is usually prefered to Scalar, as the former results in "better" representations, while Scaler is easily understandable for tests/visualisations. 

When I'll be changing the tests already, should we use RDSE as the internal encoder? Or atleast make an option to use it?

Reply:
The scalar is better for this task.
- Prefer the scalar encoder when the range of inputs is known.
- Use RDSE when the range of inputs is unknown.

Reply:
That's true, though from experience I'd say RDSE results overall in better representations. Due to the fact that it already prvodes sparse output. Theoretically SP should do all the job, but ... 
But might be a followup improvement, out of scope of the initial PR.